File: LeetcodeSwift/LeetcodeSwift/algorithms/q0107.swift
//
//  q0107.swift
//  LeetcodeSwift
//
//  Created by NowOrNever on 13/02/2020.
//  Copyright © 2020 DL. All rights reserved.
//

import Foundation

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
//107. Binary Tree Level Order Traversal II
//Easy
//
//1018
//
//186
//
//Add to List
//
//Share
//Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
//
//For example:
//Given binary tree [3,9,20,null,null,15,7],
//    3
//   / \
//  9  20
//    /  \
//   15   7
//return its bottom-up level order traversal as:
//[
//  [15,7],
//  [9,20],
//  [3]
//]
//Accepted
//275,708
//Submissions
//553,441
class q0107Solution {
    func levelOrderBottom_helper(node: TreeNode?, level: Int, result: inout [[Int]]){
        if node != nil {
            if level >= result.count {
                result.append([Int]())
            }
            result[level].append(node!.val)
            levelOrderBottom_helper(node: node?.left, level: level + 1, result: &result)
            levelOrderBottom_helper(node: node?.right, level: level + 1, result: &result)
        }
    }
    func levelOrderBottom(_ root: TreeNode?) -> [[Int]] {
        var result = [[Int]]()
        levelOrderBottom_helper(node: root, level: 0, result: &result)
        return result.reversed()
    }
}


func q0107() -> () {
    let node3 = TreeNode.init(3)
    let node9 = TreeNode.init(9)
    let node20 = TreeNode.init(20)
    let node15 = TreeNode.init(15)
    let node7 = TreeNode.init(7)
    node3.left = node9
    node3.right = node20
    node20.left = node15
    node20.right = node7
    
    let solution = q0107Solution()
    let result = solution.levelOrderBottom(node3)
    print(result)
}

File: LeetcodePython3/q0107.py
#!/usr/bin/python3
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrderBottom_helper(self, root: TreeNode, length: int, result: List[List[int]]):
        if root:
            if length > len(result):
                result.append(list())
            result[length - 1].append(root.val)
            self.levelOrderBottom_helper(root.left, length + 1, result)
            self.levelOrderBottom_helper(root.right, length + 1, result)

    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        result: List[List[int]] = list()
        self.levelOrderBottom_helper(root, 1, result)
        result.reverse()
        return result


node15 = TreeNode(15)
node7 = TreeNode(7)
node20 = TreeNode(20,node15, node7)
node9 = TreeNode(9)
node3 = TreeNode(3, node9, node20)

solu = Solution()
result = solu.levelOrderBottom(node3)
print(result)